pkg/stub/handler.go

```diff
@@ -247,3 +255,179 @@ func (h *Handler) ensureRouterDeleted(ci *ingressv1alpha1.ClusterIngress) error
 	}
 	return nil
 }
+
+// syncOperatorStatus computes the operator's current status and therefrom
+// creates or updates the ClusterOperator resource for the operator.
+func (h *Handler) syncOperatorStatus() {
+	co := &osv1.ClusterOperator{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ClusterOperator",
+			APIVersion: "operatorstatus.openshift.io/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: h.namespace,
+			// TODO Use a named constant or get name from config.
+			Name: "openshift-ingress",
+		},
+	}
+	err := sdk.Get(co)
+	isNotFound := errors.IsNotFound(err)
+	if err != nil && !isNotFound {
+		logrus.Errorf("syncOperatorStatus: error getting ClusterOperator %s/%s: %v",
+			co.Namespace, co.Name, err)
+
+		return
+	}
+
+	ns, ingresses, daemonsets, err := h.getOperatorState()
+	if err != nil {
+		logrus.Errorf("syncOperatorStatus: getOperatorState: %v", err)
+
+		return
+	}
+
+	oldCo := co
+	co = co.DeepCopy()
+	setStatusConditions(co, ns, ingresses, daemonsets)
+
+	if isNotFound {
+		co.Status.Version = operatorversion.Version
+
+		if err := sdk.Create(co); err != nil {
+			logrus.Errorf("syncOperatorStatus: failed to create ClusterOperator %s/%s: %v",
+				co.Namespace, co.Name, err)
+		} else {
+			logrus.Infof("syncOperatorStatus: created ClusterOperator %s/%s (UID %v)",
+				co.Namespace, co.Name, co.UID)
+		}
+
+		return
+	}
+
+	if clusteroperator.StatusEqual(&oldCo.Status, &co.Status) {
+		return
+	}
+
+	if err := sdk.Update(co); err != nil {
+		logrus.Errorf("syncOperatorStatus: failed to update status of ClusterOperator %s/%s: %v",
+			co.Namespace, co.Name, err)
+	}
+}
+
+// getOperatorState gets and returns the resources necessary to compute the
+// operator's current state.
+func (h *Handler) getOperatorState() (*corev1.Namespace, *ingressv1alpha1.ClusterIngressList, *appsv1.DaemonSetList, error) {
```

Comment:
How about returning plain arrays (e.g. `[]ClusterIngress`) instead? The list wrappers are cumbersome to work with and provide no additional value in this case

Reply:
Fixed.  My only concern is that I don't want to copy large structures accidentally; hopefully I've avoided doing so...